### PR TITLE
GPUCACell::find_ntuplets abort on recursion overflow

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Use the `__trap()` builtin to abort the execution from within a GPU kernel if the  `GPUCACell::find_ntuplets` recusrions overflows.

This should not be possible in Phase-1 code, and is not yet been used in Phase-2 code, but is added for extra robustness.

#### PR validation:

Running with the standard configuration does not show any regressions, not even on performance:

measurement     |   #35473     |  #35473 with `__trap()`  
----------------|-------------------|-------------------
I/O throughput  |        ~ 2 kev/s  |        ~ 2 kev/s  
quadruplets     |   1289 ±  3 ev/s  |   1292 ±  4 ev/s  
triplets        |    724 ±  3 ev/s  |    720 ±  2 ev/s  

.

Forcing the overflow to happen by artificially reducing the maximum recursion fails as expected:
```
Begin processing the 1st record. Run 1, Event 1602, LumiSection 17 on stream 0 at 05-Oct-2021 15:15:12.381 CEST
ERROR: GPUCACell::find_ntuplets reached full depth!
... (repeated 819 times)
ERROR: GPUCACell::find_ntuplets reached full depth!
terminate called after throwing an instance of 'std::runtime_error'
  what():  
/data/user/fwyzard/CMSSW_12_1_0_pre3/src/HeterogeneousCore/CUDAUtilities/src/CachingDeviceAllocator.h, line 617:
cudaCheck(error = cudaEventRecord(search_key.ready_event, search_key.associated_stream));
cudaErrorLaunchFailure: unspecified launch failure
```
followed by the usual stack trace.

### CODE REVIEW

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

File: RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h [hunk 1]
```diff
@@ -359,7 +359,11 @@ __device__ inline void GPUCACell::find_ntuplets<0>(Hits const& hh,
                                                    const unsigned int minHitsPerNtuplet,
                                                    bool startAt0) const {
   printf("ERROR: GPUCACell::find_ntuplets reached full depth!\n");
-  assert(false);  // on GPU is compiled away in NDEBUG mode
+#ifdef __GPU_ARCH__
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h:362]:
```suggestion
#ifdef __CUDA_ARCH__
```

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
assign heterogeneous

**Comment by fwyzard** [comment]
@makortel what do you think ?

**Comment by makortel** [comment]
Thanks @fwyzard. Given that there is no performance impact, I think having an explicit protection would be a little bit safer.

But this makes me wonder some questions further
* Would it be worth to provide a simple abstraction instead of spreading this fragment in all places where it could be useful? (would there be many?)
* Should we use `__trap()` to craft our own assertion function?
   * I'm honestly unsure. On one hand it would be tempting, on the other hand the standard `assert()` provides a bit more information, and our own assertion could create confusion (e.g. wrt. the message "don't use assertions in device code").

**Comment by fwyzard** [comment]
>     * Would it be worth to provide a simple abstraction instead of spreading this fragment in all places where it could be useful? (would there be many?)

I think so, but since this is the first place where we use it maybe it's not urgent.


>     * Should we use `__trap()` to craft our own assertion function?
>       
>       * I'm honestly unsure. On one hand it would be tempting, on the other hand the standard `assert()` provides a bit more information, and our own assertion could create confusion (e.g. wrt. the message "don't use assertions in device code").

We could change that message to something like use `cms::assert(...)` or `cms::cuda::assert(...)` in device code.

We can also postpone this to after we move to a portability framework, and implement something using that.

**Comment by makortel** [comment]
I'm fine with this PR as it is, the earlier questions were more for future thought (sorry for not being clear).

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by perrotta** [comment]
+1

**Pull request merged**